test/extensions/transport_sockets/alts/alts_integration_test.cc

```diff
@@ -0,0 +1,136 @@
+    handshaker_service: ")EOF" +
+                             fakeHandshakerServerAddress() + "\"";
+
+    MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), *transport_socket);
+  });
+  HttpIntegrationTest::initialize();
+  registerTestServerPorts({"http"});
+}
+
+void AltsIntegrationTest::SetUp() {
+  fake_handshaker_server_thread_ = std::make_unique<Thread::Thread>([this]() {
+    std::unique_ptr<grpc::Service> service = grpc::gcp::CreateFakeHandshakerService();
+
+    std::string server_address = Network::Test::getLoopbackAddressUrlString(version_) + ":0";
+    grpc::ServerBuilder builder;
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials(),
+                             &fake_handshaker_server_port_);
+    builder.RegisterService(service.get());
+
+    fake_handshaker_server_ = builder.BuildAndStart();
+    fake_handshaker_server_ci_.setReady();
+    fake_handshaker_server_->Wait();
+  });
+
+  fake_handshaker_server_ci_.waitReady();
+
+  const std::string yaml =
+      absl::StrCat("handshaker_service: \"", fakeHandshakerServerAddress(), "\"");
+
+  NiceMock<Server::Configuration::MockTransportSocketFactoryContext> mock_factory_ctx;
+  UpstreamAltsTransportSocketConfigFactory factory;
+
+  ProtobufTypes::MessagePtr config = factory.createEmptyConfigProto();
+  MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), *config);
+
+  ENVOY_LOG_MISC(info, "{}", config->DebugString());
+
+  client_alts_ = factory.createTransportSocketFactory(*config, mock_factory_ctx);
+}
+void AltsIntegrationTest::TearDown() {
+  HttpIntegrationTest::cleanupUpstreamAndDownstream();
+  dispatcher_->clearDeferredDeleteList();
+  if (fake_handshaker_server_ != nullptr) {
+    fake_handshaker_server_->Shutdown();
+  }
+  fake_handshaker_server_thread_->join();
+}
+
+Network::Address::InstanceConstSharedPtr getAddress(const Network::Address::IpVersion& version,
+                                                    int port) {
+  std::string url =
+      "tcp://" + Network::Test::getLoopbackAddressUrlString(version) + ":" + std::to_string(port);
+  return Network::Utility::resolveUrl(url);
+}
+
+Network::ClientConnectionPtr AltsIntegrationTest::makeAltsConnection() {
+  Network::Address::InstanceConstSharedPtr address = getAddress(version_, lookupPort("http"));
+  return dispatcher_->createClientConnection(address, Network::Address::InstanceConstSharedPtr(),
+                                             client_alts_->createTransportSocket(), nullptr);
+}
+
+INSTANTIATE_TEST_CASE_P(IpVersions, AltsIntegrationTest,
+                        testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                        TestUtility::ipTestParamsToString);
+
+TEST_P(AltsIntegrationTest, RouterRequestAndResponseWithBodyNoBuffer) {
+  ConnectionCreationFunction creator = [this]() -> Network::ClientConnectionPtr {
+    return makeAltsConnection();
+  };
+  testRouterRequestAndResponseWithBody(1024, 512, false, &creator);
+}
+
```

Comment:
I don't know how possible it is with grpc fake handshake service, but it'll be great if you can add a handshake failure test and peer validation failure test.

Reply:
Added tests to cover peer validation failure at both sides. Handshake failure case is not easy to add, which might involve changes to fake handshake server.

Reply:
One thing you might be able to do is point handshaker_service to non-exist handshaker server, that will fail handshake.

Reply:
I have added one test which set a wrong address to client's config, so that client fails to do handshake. For the server side, it is weird that Envoy always connects to the fake handshaker server even I set an arbitrary string as handshaker server's address. I am still debugging it.

Reply:
After discussing with @jiangtaoli2016 and @yihuazhang, we figure out why Envoy can still talk to fake handshaker server even the handshaker url is " ". This is because when client and Envoy are sharing one gRPC binary. When client starts talking to fake handshaker server, init_shared_resources() is called and kSharedResource->channel is created. Later when Envoy talks to fake handshaker server, it reuses the  kSharedResource->channel and continue handshake process. 

In real application use, two handshake clients (e.g. two Envoys) are depending on separate gRPC library, so this issue will not happen. 

I have removed TEST_P(AltsIntegrationTestServerWrongHandshaker, ConnectToWrongHandshakerAddress), as TEST_P(AltsIntegrationTestClientWrongHandshaker, ConnectToWrongHandshakerAddress) is sufficient to cover the error case.